test/benchmarks/netperf/nptest.go

```diff
@@ -0,0 +1,660 @@
+	if err != nil {
+		fmt.Println("Failed to append to existing file", filename, err)
+		return
+	}
+	defer fd.Close()
+
+	_, err = fd.WriteString(data)
+	if err != nil {
+		fmt.Println("Failed to append to existing file", filename, err)
+	}
+}
+
+func registerDataPoint(label string, mss int, value string, index int) {
+
+	sl, ok := dataPoints[label]
+	if ok != true {
+		dataPoints[label] = []Point{{mss: mss, bandwidth: value, index: index}}
+		dataPointKeys = append(dataPointKeys, label)
+	} else {
+		dataPoints[label] = append(sl, Point{mss: mss, bandwidth: value, index: index})
+	}
+}
+
+func flushDataPointsToCsv() {
+
+	var buffer string
+
+	// Write the MSS points for the X-axis before dumping all the testcase datapoints
+	for _, points := range dataPoints {
+		if len(points) == 1 {
+			continue
+		}
+		buffer = fmt.Sprintf("%-45s, Maximum,", "MSS")
+		for _, p := range points {
+			buffer = buffer + fmt.Sprintf(" %d,", p.mss)
+		}
+		break
+	}
+	buffer = buffer + "\n"
+	writeOutputFile(CSVFILE, buffer)
+
+	var lines []string
+	for _, label := range dataPointKeys {
+		buffer = fmt.Sprintf("%-45s,", label)
+		points := dataPoints[label]
+		var max float64
+		for _, p := range points {
+			fv, _ := strconv.ParseFloat(p.bandwidth, 64)
+			if fv > max {
+				max = fv
+			}
+		}
+
+		buffer = buffer + fmt.Sprintf("%f,", max)
+		for _, p := range points {
+			buffer = buffer + fmt.Sprintf("%s,", p.bandwidth)
+		}
+		buffer = buffer + "\n"
+		lines = append(lines, buffer)
+	}
+
+	for _, l := range lines {
+		writeOutputFile(CSVFILE, l)
+	}
+}
+
+func parseIperfTcpBandwidth(output string) string {
+	// Parses the output of iperf3 and grabs the group Mbits/sec from the output
+	match := iperf_tcp_output_regexp.FindStringSubmatch(output)
+	if match != nil && len(match) > 1 {
+		return match[1]
+	}
+	return "0"
+}
+
+func parseIperfUdpBandwidth(output string) string {
+	// Parses the output of iperf3 (UDP mode) and grabs the Mbits/sec from the output
+	match := iperf_udp_output_regexp.FindStringSubmatch(output)
+	if match != nil && len(match) > 1 {
+		return match[1]
+	}
+	return "0"
+}
+
+func parseNetperfBandwidth(output string) string {
+	// Parses the output of netperf and grabs the Bbits/sec from the output
+	match := netperf_output_regexp.FindStringSubmatch(output)
+	if match != nil && len(match) > 1 {
+		return match[1]
+	}
+	return "0"
+}
+
+func (t *NetPerfRpc) ReceiveOutput(data *WorkerOutput, reply *int) error {
+	global_lock.Lock()
+	defer global_lock.Unlock()
+
+	testcase := testcases[currentJobIndex]
+
+	var outputLog string
+	var bw string
+
+	switch data.Type {
+	case IPERF_TCP_TEST:
+		mss := testcases[currentJobIndex].MSS - MSS_STEP_SIZE
+		outputLog = outputLog + fmt.Sprintln("Received TCP output from worker", data.Worker, "for test", testcase.Label,
+			"from", testcase.SourceNode, "to", testcase.DestinationNode, "MSS:", mss, "\n") + data.Output
+		go writeOutputFile(OUTPUT_CAPTURE_FILE, outputLog)
```

Comment:
I was not checking the logic so I don't know if this is an error or not, but it's certainly a smell. You start a goroutine under the lock and don't wait for it to finish before releasing the lock. It may be OK, but you may end up with some unfinished goroutines when program ends.